### PR TITLE
[ALLUXIO-2522] Fix free operation to respect "pinned" and "persistenceState"

### PR DESCRIPTION
**Pull request by apc999**
https://alluxio.atlassian.net/browse/ALLUXIO-2522

### CODE REVIEW

**Comment by apc999** [comment]
Jenkins, test this please

**Comment by apc999** [comment]
@calvinjia Document udpated

File: core/server/src/main/java/alluxio/master/file/FileSystemMaster.java [hunk 1]
```diff
@@ -2075,63 +2077,97 @@ private long journalPersistedInodes(List<Inode<?>> persistedInodes) {
...
-
+      TempInodePathForDescendant tempInodePath = new TempInodePathForDescendant(inodePath);
       // We go through each inode.
       for (int i = freeInodes.size() - 1; i >= 0; i--) {
         Inode<?> freeInode = freeInodes.get(i);
 
         if (freeInode.isFile()) {
+          if (freeInode.getPersistenceState() != PersistenceState.PERSISTED) {
+            throw new UnexpectedAlluxioException(ExceptionMessage.CANNOT_FREE_NON_PERSISTED_FILE
+                .getMessage(mInodeTree.getPath(freeInode)));
```
Comment (apc999) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: force will help unpin files before free, not persist files before free---unpin is a metadata op but persist is not

File: tests/src/test/java/alluxio/master/file/FileSystemMasterClientRestApiTest.java [hunk 2]
```diff
@@ -141,7 +142,9 @@ public void getStatus() throws Exception {
     Assert.assertEquals(0, fileInfo.getLength());
   }
 
+  // TODO(binfan): re-enable this test
   @Test
+  @Ignore
```
Comment (apc999) [inline tests/src/test/java/alluxio/master/file/FileSystemMasterClientRestApiTest.java]: it requires to update the FSMaster Rest API which is deprecated already. I will make a separate PR on that

File: core/server/src/main/java/alluxio/master/file/FileSystemMaster.java [hunk 3]
```diff
@@ -2689,18 +2725,24 @@ private boolean unmountInternal(AlluxioURI uri) throws InvalidPathException {
...
    * @throws AccessControlException if permission checking fails
    * @throws InvalidPathException if the path is invalid for the id of the file
+   * @throws UnexpectedAlluxioException if the file or directory can not be freed
    */
   // Currently used by Lineage Master
   // TODO(binfan): Add permission checking for internal APIs
   public void resetFile(long fileId)
-      throws FileDoesNotExistException, InvalidPathException, AccessControlException {
+      throws UnexpectedAlluxioException, FileDoesNotExistException, InvalidPathException,
+      AccessControlException {
```
Comment (apc999) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: this is after formatter

File: core/server/src/main/java/alluxio/master/file/FileSystemMaster.java [hunk 4]
```diff
@@ -3059,7 +3101,7 @@ public void heartbeat() {
               LOG.debug("File {} is expired. Performing action {}", inode.getName(), ttlAction);
               switch (ttlAction) {
                 case FREE:
-                  free(path, true);
+                  free(path, FreeOptions.defaults().setForced(true));
```
Comment (apc999) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: done

File: core/server/src/main/java/alluxio/master/file/FileSystemMasterClientServiceHandler.java [hunk 5]
```diff
@@ -125,11 +127,20 @@ public Void call() throws AlluxioException, IOException {
...
+  public void free(final String path, final boolean recursive, final FreeTOptions options)
+      throws AlluxioTException {
     RpcUtils.call(new RpcCallable<Void>() {
       @Override
       public Void call() throws AlluxioException {
-        mFileSystemMaster.free(new AlluxioURI(path), recursive);
+        if (options == null) {
+          // For Alluxio client v1.4 or earlier.
+          // NOTE, we try to be conservative here so early Alluxio clients will not be able to force
+          // freeing pinned items but see the error thrown.
```
Comment (apc999) [inline core/server/src/main/java/alluxio/master/file/FileSystemMasterClientServiceHandler.java]: I and @jsimsa  had a discussion on this and Jiri convinced me freeing pinned files silently for older client without telling users is bad. So the old client will experience different free result on pinned files when it talks to new server (which is anyway a corner case IMO). Let me know if you think differently.

**Comment by apc999** [comment]
@calvinjia  comments addressed or filled in with replies. PTAL. thanks

File: core/server/src/main/java/alluxio/master/file/FileSystemMaster.java (hunk 1 above)
Comment (jsimsa) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: I guess Calvin's point is that traditionally, the semantics of "force" is to try to complete the operation no matter what.

File: tests/src/test/java/alluxio/master/file/FileSystemMasterClientRestApiTest.java (hunk 2 above)
Comment (jsimsa) [inline tests/src/test/java/alluxio/master/file/FileSystemMasterClientRestApiTest.java]: agreed, re-enabling this test is low priority in my mind as it is testing API that is deprecated

File: core/server/src/main/java/alluxio/master/file/FileSystemMaster.java (hunk 3 above)
Comment (apc999) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: @jsimsa , yes

Comment (jsimsa) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: ack

**Pull request merged**